### PR TITLE
Feature/GAL-851

### PR DESCRIPTION
**Pull request by MichaelW-NHS**
## Description

https://nhsd-jira.digital.nhs.uk/browse/GAL-851

Receipt of Result: Validate Test Result Report (TRR) using FHIR validation service (Step 1)

## Type of changes

- [ ] Refactoring (non-breaking change)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

**Review by chenglawson1 — APPROVED** [review]

File: terraform/src/testResultReportFhirValidationLambda/testResultReportFhirValidationLambda.js
Comment (Amar-NHS) [inline terraform/src/testResultReportFhirValidationLambda/testResultReportFhirValidationLambda.js:1]: might have been missed but for future reference, there were existing helper function to perform S3 actions - "deleteObjectFromS3", "pushJsonToS3". They do same thing as "deleteTRRinS3Bucket" & "putTRRInS3Bucket". I believe we were planning to move them in lambda layers as common helper functions.

Comment (MichaelW-NHS) [reply terraform/src/testResultReportFhirValidationLambda/testResultReportFhirValidationLambda.js:1]: Are those currently existing? I see those two functions but not as helper function, just defined within a lambda.

File: terraform/src/testResultReportFhirValidationLambda/testResultReportFhirValidationLambda.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+export async function putTRRInS3Bucket(testResultReport, reportName, bucketName, client) {
+  try {
+    const response = await client.send(
+      new PutObjectCommand({
+        Bucket: `${ENVIRONMENT}-${bucketName}`,
+        Key: reportName,
+        Body: JSON.stringify(testResultReport),
+      })
+    );
+    console.log(`Successfully pushed to ${bucketName}/${reportName}`);
```
[the remark this answers is not in the document]
Comment (MichaelW-NHS) [reply terraform/src/testResultReportFhirValidationLambda/testResultReportFhirValidationLambda.js:84]: It's being pushed to a different bucket.

File: terraform/main.tf [hunk 2]
```diff
@@ -1576,6 +1576,43 @@ module "caas_feed_delete_records_lambda_cloudwatch" {
...
+  bucket_id     = module.proccessed_nrds.bucket_id
+  bucket_arn    = module.proccessed_nrds.bucket_arn
+  lambda_arn    = module.test_result_report_fhir_validation_lambda.lambda_arn
+  filter_prefix = "record_"
+}
+
+data "aws_secretsmanager_secret_version" "fhir_validation_service_url" {
+  secret_id = "FHIR_VALIDATION_SERVICE_URL"
+}
+
```
[the remark this answers is not in the document]
Comment (MichaelW-NHS) [reply terraform/main.tf]: Removed this and kept the getSecret function instead.

File: terraform/src/testResultReportFhirValidationLambda/testResultReportFhirValidationLambda.js
Comment (Amar-NHS) [reply terraform/src/testResultReportFhirValidationLambda/testResultReportFhirValidationLambda.js:1]: Yh they are existing but not labelled as helper functions or inside a helper directory yet.

File: terraform/src/testResultReportFhirValidationLambda/testResultReportFhirValidationLambda.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Amar-NHS) [reply terraform/src/testResultReportFhirValidationLambda/testResultReportFhirValidationLambda.js:84]: I see. Reading the log sort of indicates that the report is being pushed in a sub-dir inside the bucket name

**Review by Amar-NHS — APPROVED** [review]

**Review by chenglawson1 — APPROVED** [review]

**Pull request merged**